proposals/NNNN-fix-expressiblebystringinterpolation.md

```diff
@@ -0,0 +1,471 @@
+
+We propose a set of `String.StringInterpolation.appendInterpolation` overloads which exactly match the current `init(stringInterpolationSegment:)` overloads, so “normal” interpolations will work exactly as before.
+
+“Strange” interpolations like `\(x, y)` or `\(foo: x)` which are currently accepted by the Swift compiler will be errors in Swift 5 mode. In Swift 4.2 mode, we will preserve the existing behavior with a warning; this means that Swift 4.2 code will only be able to use `appendInterpolation` overloads with a single unlabeled parameter, unless all other parameters have default values. Migration involves inserting an extra pair of parens or removing an argument label to preserve behavior.
+
+## Effect on ABI stability
+
+`ExpressibleByStringInterpolation` will need to be ABI-stable starting in Swift 5; we should adopt this proposal or some alternative and un-deprecate `ExpressibleByStringInterpolation` before that.
+
+## Effect on API resilience
+
+This API is pretty foundational and it would be difficult to change compatibly in the future.
+
+## Alternatives considered
+
+### Variadic-based designs
+
+We considered several designs which, like the current design, passed segments to a variadic parameter. For example, we could wrap literal segments in `init(stringLiteral:)` instead of `init(stringInterpolationSegment:)` and otherwise keep the existing design:
+
+```swift
+String(stringInterpolation:
+  String(stringLiteral: "hello "),
+  String(stringInterpolationSegment: name),
+  String(stringLiteral: "!"))
+```
+
+Or we could use an enum to differentiate literal segments from interpolated ones:
+
+```swift
+String(stringInterpolation:
+  .literal("hello "),
+  .interpolation(String.StringInterpolationType(name)),
+  .literal("!"))
+```
+
+However, this requires that conformers expose a homogenous return value, which has expressibility and/or efficiency drawbacks. The proposed approach, which is statement based, keeps this as a detail internal to the conformer.
+
+### Have a formal `appendInterpolation(_:)` requirement
+
+We considered having a formal `appendInterpolation(_:)` requirement with an unconstrained generic parameter to mimic current behavior. We could even have a default implementation that vends strings and still honor overloading.
+
+However, we would have to give up on conformers being able to restrict the types or interpolation segment forms permitted.
+
+## Future directions
```

Comment:
I wonder if this section is better entitled "Potential Uses"? And maybe hoisting higher into the motivation section. It's basically a cookbook of neat stuff this proposal enables, rather than ways we could further enhance interpolation itself.

Reply:
Michael suggested the same thing privately; I'll probably do that.